### ISSUE TITLE
make a 4.5.1 maintenance release

### ISSUE THREAD

**Issue by sezero** [libxmp/libxmp#381]
E.g. #380 was a fix for 4.5.0-introduced xmp_set_row().

**Comment by AliceLR** [comment]
Not sure that there are any other urgent/crash/vulnerability/etc. bugs that are open currently.

The person who reported #380 to me also indicated there might be a bug where using `xmp_set_position` for some modules might have problems, but this might be specific to MegaZeux, which nukes sequences since they break more than they help in that specific use case. If I can get a test file from this person I'll diagnose this potential issue and follow up. (I actually received the report for #380 prior to the 4.5.0 release, I just forgot about it since it was a DM instead of being submitted to an issue tracker... oops.)

**Comment by AliceLR** [comment]
I now have a copy of the module in question and it does look like a libxmp bug: when there are more than 16 sequences, all orders after sequence 16 are marked as having sequence `ff`, which makes it impossible to jump to them with `xmp_set_position`. It would be nice for there to be a way to turn off sequence scanning beyond anything required to detect repeats in the main sequence and allow jumping to any order regardless of the sequence count (like most other players allow).

**Comment by sezero** [comment]
https://github.com/libxmp/libxmp/issues/385 is a major problem.

**Comment by AliceLR** [comment]
We're up to [nearly 100 patches linked to this milestone](https://github.com/libxmp/libxmp/milestone/4) now, with no end in sight... should we draw some lines for when to stop adding features so we can fuzz and release this thing? : D

**Comment by sezero** [comment]
>     * Adapting my permissive licensed [ARC](https://github.com/AliceLR/megazeuxtests/blob/master/src/dimgutil/ArcFS.cpp) [and](https://github.com/AliceLR/megazeuxtests/blob/master/src/dimgutil/SparkFS.cpp) [ArcFS](https://github.com/AliceLR/megazeuxtests/blob/master/src/dimgutil/arc_unpack.c) [depackers](https://github.com/AliceLR/megazeuxtests/blob/master/src/dimgutil/arc_crc16.c) to libxmp.

This would be _really_ nice to have.

**Comment by AliceLR** [comment]
Most of the stuff listed above is done, the main stuff left I think would be nice is:

* Compile time flags to select different Vorbis libraries.
* There's a new pending ASan patch I have a couple of fixes left for: https://github.com/libxmp/libxmp/compare/master...AliceLR:fuzz-patch-20220217
* Further ASan and other sanitizer testing/fuzzing due to the two big patches that were just merged.

There's also #387, which *looks* pretty close on paper, but I don't know how long actually completing it will take. LHA seems like a pain. Has Claudio said anything about gunzip?

**Comment by sezero** [comment]
There is also this: https://github.com/richgel999/miniz/pull/159#issuecomment-1046220324

>  Has Claudio said anything about gunzip?

Nope.

**Comment by sezero** [comment]
>   There's a new pending ASan patch I have a couple of fixes left for: [master...AliceLR:fuzz-patch-20220217](https://github.com/libxmp/libxmp/compare/master...AliceLR:fuzz-patch-20220217)

> There is also this: [richgel999/miniz#159 (comment)](https://github.com/richgel999/miniz/pull/159#issuecomment-1046220324)

Here is a patch for that (in `git format-patch` format) for you to include in your patchset: [159_patch.txt](https://github.com/libxmp/libxmp/files/8132129/159_patch.txt)

**Comment by sezero** [comment]
> > Here is a patch for that
> 
> Will apply when I get done fixing the remaining bugs found working on that branch, thanks.

Mainstream applied the patch in this form:  https://github.com/richgel999/miniz/commit/9eedb97176408008a2263cef9b4bac5c7f955d4f

**Comment by AliceLR** [comment]
Trying to finish up another project that's close to done, then I'll look at the Vorbis thing. As far as fuzzing the rest of libxmp, I *think* it might be in a good place. For all three sanitizers, libFuzzer was pretty slow at finding new issues (nothing new for over a week in the case of UBSan).

**Comment by sezero** [comment]
>  then I'll look at the Vorbis thing.

The `Vorbis thing`? What is left of it?

**Comment by sezero** [comment]
IMO, these are the possible issues we should resolve before  a new release:
- https://github.com/libxmp/libxmp/issues/506
- https://github.com/libxmp/libxmp/issues/505
- https://github.com/libxmp/libxmp/issues/497 (resolved already?)
- Maybe this? https://github.com/libxmp/libxmp/issues/578
- Maybe this? https://github.com/libxmp/libxmp/issues/499
- Maybe this? (all there is left is unlha and gunzip) https://github.com/libxmp/libxmp/issues/481
- This should be closed already, IMO: https://github.com/libxmp/libxmp/issues/481
- Maybe this, only if there is existing work for it: https://github.com/libxmp/libxmp/issues/379
- Maybe this? (not easy work, though) https://github.com/libxmp/libxmp/issues/574

**Comment by AliceLR** [comment]
Working libvorbis (tested)/Tremor (untested) compatibility layer: https://gist.github.com/AliceLR/6c64ebb0efafbd12ee2682eaf1b99f3a

After looking at the Makefiles I'm not sure I really want to implement the parts to selectively disable vorbis.c right now. On top of that, getting the autoconf bits right seems tricky enough it might be worth just delaying this one (I'll hack it into the MegaZeux fork of libxmp to make sure it gets *some* testing, at least).

I also found an unrelated OXM bug that I'll submit a patch for shortly.

> IMO, these are the possible issues we should resolve before a new release:

I'll go through these in a bit, thanks. Some of them stick out as being unfeasible though (#505 and #506).

**Comment by sezero** [comment]
>  edit: misremembered the affected module for #505, that temporary "quirk" hack would be easy to add.

Posted an initial working solution for 506 to #506

**Comment by AliceLR** [comment]
> * #505

Yep, I was completely wrong about this one. Even an initial actual patch wasn't too bad to put together, and it's probably not *that* much of a stretch to adapt it into part of the scan (not sure if I want to yet). I need many more modules to test this with, but determining whether or not this works with them is pretty easy, and this patch will only ever affect one format.

> * #497 (resolved already?)

To go through the ones that weren't fixed:
* `mod/PortaSmpChange.mod` - IMO this can wait. `read_event_mod` touches approximately a bazillion formats...
* `it/s7xinsnum.it` - when I was doing all of the *other* IT fixes I looked into this as well, but came up with nothing. It's still broken. Since it's just IT affected it's at least reasonable to look into.
* "When toneporta is combined with a sample change [...]" - same.

> * Maybe this? #578

The patch I posted here is bad and not worth applying. A real fix might require redoing how envelopes are read (non-trivial), and it's shared code between several formats.

> * Maybe this? #499

Not a bug... I went ahead and just closed this one.

> * Maybe this? (all there is left is unlha and gunzip) #387

The earlier discussion in this issue still seems applicable...

> * This should be closed already, IMO: #481

I'll go ahead and submit the last couple of patches soon so it can finally be closed...

> * Maybe this, only if there is existing work for it: #379

I haven't touched these at all.

> * Maybe this? (not easy work, though) #574

I don't think this one is happening any time soon... 😅

**Comment by sezero** [comment]
I bumped version to 4.5.1 and updated the changelogs as of a4b0b3c07bb682cd226bfaac8da570f9c0336dbe
Corrections would be welcome.

**Comment by sagamusix** [comment]
What's the status of this? At this point it's hardly a maintenance release I guess...

**Comment by sezero** [comment]
> What's the status of this? At this point it's hardly a maintenance release I guess...

No, definitely not a maintenance release, especially after https://github.com/libxmp/libxmp/commit/32e8f08d6ac059b6e288b0b381a2fb6f12e82465 (PR/#641).

As for what's left: As far as I can tell, the ones listed at https://github.com/libxmp/libxmp/issues/381#issuecomment-1096177872, particularly bug #505/#619 which is waiting for @AliceLR to finalize. I also want to make the library fully MIT (GPL-free) licensed which involves replacing the lha depacker (using bits from https://github.com/fragglet/lhasa might be a solution but no one has got to it yet.)

**Comment by sagamusix** [comment]
I really don't think it makes  much sense to wait for a fix for long-standing playback issues. If it was a regression in 4.5.0, sure, but neither of those issues are new.

**Comment by AliceLR** [comment]
I'd argue #505 is worth blocking for if only because it's a serious playback bug and the patch is more or less finished; I just need to add new MD5s and do some cleanup. There's a couple of other loose ends I'd like to get in: xmp.h defines for stereo samples (trivial), a patch for MED SS 2.x default pitch events (trivial, semi-regression), MPT 1.16 XM detection (already mostly done). I'll try to get these in tonight, sorry...

I haven't got to these yet because I've had my hands full rebuilding a server and writing new scripts and phpBB plugins for a forum migration. The server in question is stuck on a dead and potentially unsecure version of PHP until I get this done, so it's been getting most of my attention. The old forum is IP.Board, hence how long this has been taking.

I don't really consider the rest worth blocking over; a better fix would be to tag release versions far more frequently than we currently do, so the future patches for those don't get tied up like this release has been.

**Comment by AliceLR** [comment]
Since opening #653 I've been running fuzzing jobs again. So far they've found yet another The Player issue, but nothing else.

**Comment by AliceLR** [comment]
Is this ready to tag?

**Comment by sezero** [comment]
> Is this ready to tag?

Fully MIT (GPL-free) licensed library was _really_ something I wanted to see happen (which involves replacing the lha depacker -- using bits from https://github.com/fragglet/lhasa maybe?), but but still haven't done anything about it.

Maybe tag without it?

**Comment by AliceLR** [comment]
> Maybe tag without it?

I think that would be best, yes. I don't know when I'll have the mental fortitude for another depacker replacement. (I think as long 4.6.1 or 4.7.0 or whatever doesn't take 2 years, it'll be fine.)

**Comment by sezero** [comment]
Tagged and created a release: https://github.com/libxmp/libxmp/releases/tag/libxmp-4.6.0

@cmatsuoka: The tarballs must be mirrored at sourceforge

**Comment by AliceLR** [comment]
Funny enough, 3 defects just showed up in my fuzzing this afternoon after 2 weeks of absolutely nothing 🤦‍♀️

I've been testing various worker counts and compile flags for fuzzing on my 2GB Le Potato and, while slow, it seems plausible as a low-power and low-heat fuzzer that can run continuously and maybe find some of these bugs faster.

**Comment by sezero** [comment]
> Funny enough, 3 defects just showed up in my fuzzing this afternoon after 2 weeks of absolutely nothing woman_facepalming

Unavoidable -- will be subject for 4.6.1 or whatever.

Where are the defects? Do you have patches?

**Comment by sezero** [comment]
> Tagged and created a release: https://github.com/libxmp/libxmp/releases/tag/libxmp-4.6.0

BTW: With the way I attached the tarballs, there are no predictable download links for them. Should I edit the release and add the tarballs under `Assets` thing?

**Comment by AliceLR** [comment]
> Where are the defects? Do you have patches?

Not yet, I intend to start on them soon though.

1) There's a previously known defect in ProWizard, especially affecting the CRB depacker (heatseek.c), where `PW_REQUEST_DATA` is provided very small changes, causing the test function to be called a large number of times with very little progress between each call. I have a couple of fixes planned for this.
2) Undiagnosed issue in hmn_extras.c.
3) Undiagnosed issue in stb_vorbis that looked quite bad.

> > Tagged and created a release: https://github.com/libxmp/libxmp/releases/tag/libxmp-4.6.0
> 
> BTW: With the way I attached the tarballs, there are no predictable download links for them. Should I edit the release and add the tarballs under `Assets` thing?

This is typically what I do with MegaZeux.

**Comment by sezero** [comment]
> > BTW: With the way I attached the tarballs, there are no predictable download links for them. Should I edit the release and add the tarballs under `Assets` thing?
> 
> This is typically what I do with MegaZeux.

OK, edited the release.

**Comment by AliceLR** [comment]
> 1. There's a previously known defect in ProWizard, especially affecting the CRB depacker (heatseek.c), where `PW_REQUEST_DATA` is provided very small changes, causing the test function to be called a large number of times with very little progress between each call. I have a couple of fixes planned for this.

I might have just revealed a bunch of underlying out-of-bounds read bugs in the ProWizard test functions by doing this please be patient :(
